tcms/testplans/tests/test_views.py

```diff
@@ -54,3 +55,125 @@ def test_view_tags_without_permissions(self):
         # assert tag actions are shown
         self.assertNotContains(response, 'Add Tag')
         self.assertContains(response, '<span class="disabled grey">%s</span>' % _('Remove'))
+
+
+class EditPlanViewTest(LoggedInTestCase):
+    test_plan_name = 'plan name'
+
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+
+        user_should_have_perm(cls.tester, perm='testplans.change_testplan')
+        cls.product = ProductFactory()
+        cls.product_version = VersionFactory(product=cls.product)
+        cls.test_plan_type = PlanTypeFactory()
+        cls.test_plan_1 = TestPlanFactory()
+        cls.testplan_edit_url = reverse('plan-edit', args=[cls.test_plan_1.pk])
+
+        cls.testplan_edit_data = {
+            'author': cls.tester.pk,
+            'product': cls.product.pk,
+            'product_version': cls.product_version.pk,
+            'type': cls.test_plan_type.pk,
+            'name': cls.test_plan_1.name,
+
+            'email_settings-0-auto_to_plan_author': 'off',
+            'email_settings-0-auto_to_case_owner': 'off',
+            'email_settings-0-auto_to_case_default_tester': 'off',
+            'email_settings-0-notify_on_case_update': 'off',
+            'email_settings-0-notify_on_plan_update': 'off',
+
+            'email_settings-0-id': cls.test_plan_1.emailing.pk,
+            'email_settings-TOTAL_FORMS': '1',
+            'email_settings-INITIAL_FORMS': '1',
+            'email_settings-MIN_NUM_FORMS': '0',
+            'email_settings-MAX_NUM_FORMS': '1',
+            'is_active': True,
+        }
+
+    def test_show_testplan_edit_page(self):
+        response = self.client.get(self.testplan_edit_url)
+        self.assertContains(response, _('Edit TestPlan'))
+        self.assert_notify_form(response)
+        self.assertContains(
+            response,
+            '<input class="bootstrap-switch" name="is_active" type="checkbox" checked>',
+            html=True)
+
+    def test_edit_testplan_text_field(self):
+        edit_data = self.testplan_edit_data.copy()
+        new_text = 'Edited: {0}'.format(self.test_plan_1.text)
+        edit_data['text'] = new_text
+
+        response = self.client.post(self.testplan_edit_url, data=edit_data, follow=True)
+
+        self.assertContains(response, new_text, html=True)
+        self.test_plan_1.refresh_from_db()
+        self.assertEqual(new_text, self.test_plan_1.text)
+
+    # Issue: When there is an invalid value on notify_formset,
+    # returned response is always the default (ON).
+    def test_with_invalid_notify_form_value(self):
+        test_plan = TestPlanFactory(
+            author=self.tester)
```

Comment:
I don't see the reason why you should use a different TP for this particular scenario. Using the existing one will work equally well.

Reply:
See reply;
https://github.com/kiwitcms/Kiwi/pull/1677#discussion_r429549452

Reply:
This comment is bonkers. There is absolutely no difference for the code path whether you have 1 or 2 or a 1000 objects in the database.  When the notify_formset is invalid it is passed directly to get_context_data() and then render_to_response(). 

This test should use the existing self.test_plan_1 variable as well.

Reply:
I did check it and your explanation is incorrect. 

Add `print('++++ form_valid - notify_formset is NOT valid', notify_formset.errors)` in form_valid(), after the if condtition which checks notify_formset and you will get:

```
++++ form_valid - notify_formset is NOT valid [{'plan': ['Test plan email settings with this Plan already exists.']}]
```

The reason for this error is that `'email_settings-0-id': cls.test_plan_1.emailing.pk` points to the previous TP (let's call it TP-1) but you are trying to edit TP-2. The value `invalid_value` for the ON/OFF button doesn't have any effect in this case.

Which is strange but I've commented before that I've seen this behavior, https://github.com/kiwitcms/Kiwi/pull/1677#discussion_r432384387. Let me debug some more to see what is happening underneath Django. If we can't reliably set notify_formset fields to invalid values then there isn't much point in trying to implement this whole scenario.

Reply:
Building on the previous comment:


```
class BooleanField(Field):
    def to_python(self, value):
        if isinstance(value, str) and value.lower() in ('false', '0'):
            value = False
        else:
            value = bool(value)
```

bool('invalid') -> True - this is the definition of how strings are evaluated in a boolean context in the Python language. The underlying form field relies on this conversion (with a few special cases b/c non empty strings are always True) and doesn't perform any other validation because it knows that it will get either True or False.

So it is not possible to make the notify_formset invalid by manipulating any of these boolean fields. 

You can however set `email_settings-0-id` to a different ID/non-existing ID and the formset will become invalid. This can also be seen in the browser, just add `value="-5"` for example in the HTML before pressing the Save button on this page.

With the current patch in views.py the page is reloaded and any buttons which have been manipulated (e.g. they were ON by default but were turned OFF by the user) are still preserved. 

That is the only way to test the path which deals with invalid notify_form. 

So the current implementation of this scenario is mostly on track but that seems to be by accident. Needs small adjustments.

Reply:
Ahh ok, sorry for the inconvenience, that was my fault, i didn't dive into details. :)